### PR TITLE
Make type information of NewArray more precise

### LINKED ISSUE #1776: [Bug]: Incorrect variable type for NewArray operation
### Describe the issue:

When using `new` expression to initialize a dynamic array, e.g., `uint[] memory arr = new uint[](2);`, slither convert it to a `NewArray` IR operation. 
However, the type of `lvalue`, `arr` variable in the example, is incorrectly set as `uint256` instead of `uint256[]`. 

### Code example to reproduce the issue:

```solidity
contract A {
    function foo() public returns (uint) {
        uint[] memory arr = new uint[](2);
        return 0;
    }
}
```

### Version:

0.9.2

### Relevant log output:

```shell
Contract A
        Function A.foo() (*)
                Expression: arr = new uint256[](2)
                IRs:
                        TMP_1 = new uint256[](2)
                        arr(uint256[]) := TMP_1(uint256) # TMP_1 has wrong type
                Expression: 0
                IRs:
                        RETURN 0
```

### PR DESCRIPTION
**Pull request by Troublor**
Fix #1776, i.e., incorrect `array_type` in `NewArray` operation. 
This PR also makes the `array_type` field more precise by including the length of the fixed-sized nested array.

### CODE REVIEW

File: tests/test_ssa_generation.py [hunk 1]
```diff
@@ -1077,3 +1079,28 @@ def test_issue_1748():
...
+        new_op = operations[0]
+        lvalue = new_op.lvalue
+        lvalue_type = lvalue.type
+        assert isinstance(lvalue_type, ArrayType)
+        lvalue_type1 = lvalue_type.type
+        assert isinstance(lvalue_type1, ArrayType)
+        assert lvalue_type1.length_value.value == "10"
+        lvalue_type2 = lvalue_type1.type
+        assert isinstance(lvalue_type2, ArrayType)
+        assert lvalue_type2.length_value.value == "5"
```
Comment (0xalpharush) [inline tests/test_ssa_generation.py:1106]: This seems backwards?

Comment (Troublor) [reply tests/test_ssa_generation.py:1106]: What do you mean `backwards`? 
Here, I am checking the nested array type and length should be parsed and converted to IR correctly.

Comment (0xalpharush) [reply tests/test_ssa_generation.py:1106]: Nvm, I misunderstood which length value corresponded to which type. This seems correct

File: tests/test_ssa_generation.py [hunk 2]
```diff
@@ -1077,3 +1079,28 @@ def test_issue_1748():
...
+    """
+    with slither_from_source(source) as slither:
+        c = slither.get_contract_from_name("Contract")[0]
+        f = c.functions[0]
+        operations = f.slithir_operations
+        new_op = operations[0]
+        lvalue = new_op.lvalue
+        lvalue_type = lvalue.type
+        assert isinstance(lvalue_type, ArrayType)
+        lvalue_type1 = lvalue_type.type
```
Comment (0xalpharush) [inline tests/test_ssa_generation.py:1101]:
```suggestion
        lvalue_type1 = lvalue_type.type
        assert lvalue_type1.is_dynamic
```
We can add the assertion that this array is dynamically sized and perhaps the inverse for those statically sized

Comment (Troublor) [reply tests/test_ssa_generation.py:1102]: It seems to me that only lvalue_type should be dynamic.
lvalue_type1 and lvalue_type2 are statically sized.
I have updated the test according to your suggestion in 4768a08.

**Comment by 0xalpharush** [comment]
I think this makes sense but it will need @montyly's review too. Would you make this change against `dev` if possible please?

**Comment by Troublor** [comment]
> I think this makes sense but it will need @montyly's review too. Would you make this change against `dev` if possible please?

The base branch of this PR has been made to `dev` branch.

**Review by 0xalpharush — APPROVED** [review]

**Comment by 0xalpharush** [comment]
@Troublor Do you mind resolving the merge conflicts please?

**Comment by montyly** [comment]
This is a great improvement, thanks @Troublor

**Pull request merged**